safehttp/plugins/csp/csp.go

```diff
@@ -0,0 +1,201 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package csp
+
+import (
+	"context"
+	"crypto/rand"
+	"encoding/base64"
+	"errors"
+	"fmt"
+	"strings"
+
+	"github.com/google/go-safeweb/safehttp"
+)
+
+var randReader = rand.Reader
+
+// nonceSize is the size of the nonces in bytes. According to the CSP3 spec it should
+// be larger than 16 bytes. 20 bytes was picked to be future proof.
+// https://www.w3.org/TR/CSP3/#security-nonces
+const nonceSize = 20
+
+func generateNonce() string {
+	b := make([]byte, nonceSize)
+	_, err := randReader.Read(b)
+	if err != nil {
+		panic(fmt.Errorf("failed to generate entropy using crypto/rand/RandReader: %v", err))
+	}
+	return base64.StdEncoding.EncodeToString(b)
+}
+
+// Policy defines a CSP policy.
+type Policy struct {
+	// serialize serializes this policy for use in a Content-Security-Policy header
+	// or in a Content-Security-Policy-Report-Only header. A nonce will be provided
+	// to serialize which can be used in 'nonce-{random-nonce}' values in directives.
+	serialize func(nonce string) string
+}
+
+type ctxKey struct{}
+
+// Nonce retrieves the nonce from the given context. If there is no nonce stored
+// in the context, an error will be returned.
+func Nonce(ctx context.Context) (string, error) {
+	v := ctx.Value(ctxKey{})
+	if v == nil {
+		return "", errors.New("no nonce in context")
+	}
+	return v.(string), nil
+}
+
+// StrictCSPBuilder can be used to build a strict, nonce-based CSP.
+//
+// See https://csp.withgoogle.com/docs/strict-csp.html for more info.
+type StrictCSPBuilder struct {
+	// StrictDynamic controls whether script-src should contain the 'strict-dynamic'
+	// value.
+	//
+	// See https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Security-Policy/script-src#strict-dynamic
+	// for more info.
+	StrictDynamic bool
```

Comment:
Maybe we should have a `NoStrictDynamic`. Strict dynamic as default seems reasonable to me TBH.